llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td

```diff
@@ -838,7 +890,201 @@ def : InstAlias<"cfsd $rs2, (${rs1})", (CFSD FPR64:$rs2, GPCR:$rs1, 0), 0>;
+  let Inst{10} = imm{8};
+  let Inst{6-5} = imm{7-6};
+}
+
+def C_CLW : CCheriLoad_ri<0b010, "c.clw", GPRC, uimm7_lsb00> {
+  bits<7> imm;
+  let Inst{12-10} = imm{5-3};
+  let Inst{6} = imm{2};
+  let Inst{5} = imm{6};
+}
+
+let DecoderNamespace = "RISCV32CapModeOnly_",
+    Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV32, IsCapMode] in
+def C_CLC_64 : CCheriLoad_ri<0b011, "c.clc", GPCRC, uimm8_lsb000> {
+  bits<8> imm;
+  let Inst{12-10} = imm{5-3};
+  let Inst{6-5} = imm{7-6};
+}
+
+let Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV64, IsCapMode] in
+def C_CLD : CCheriLoad_ri<0b011, "c.cld", GPRC, uimm8_lsb000> {
+  bits<8> imm;
+  let Inst{12-10} = imm{5-3};
+  let Inst{6-5} = imm{7-6};
+}
+
+let DecoderNamespace = "RISCV32CapModeOnly_",
+    Predicates = [HasCheri, HasCheriRVC, HasStdExtC, HasStdExtD, IsRV32, IsCapMode] in
+def C_CFSD : CCheriStore_rri<0b101, "c.cfsd", FPR64C, uimm8_lsb000> {
+  bits<8> imm;
+  let Inst{12-10} = imm{5-3};
+  let Inst{6-5} = imm{7-6};
+}
+
+let Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV64, IsCapMode] in
+def C_CSC_128 : CCheriStore_rri<0b101, "c.csc", GPCRC, uimm9_lsb0000> {
+  bits<9> imm;
+  let Inst{12-11} = imm{5-4};
+  let Inst{10} = imm{8};
+  let Inst{6-5} = imm{7-6};
+}
+
+def C_CSW : CCheriStore_rri<0b110, "c.csw", GPRC, uimm7_lsb00> {
+  bits<7> imm;
+  let Inst{12-10} = imm{5-3};
+  let Inst{6} = imm{2};
+  let Inst{5} = imm{6};
+}
+
+let DecoderNamespace = "RISCV32CapModeOnly_",
+    Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV32, IsCapMode]  in
+def C_CSC_64 : CCheriStore_rri<0b111, "c.csc", GPCRC, uimm8_lsb000> {
+  bits<8> imm;
+  let Inst{12-10} = imm{5-3};
+  let Inst{6-5} = imm{7-6};
+}
+
+let Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV64, IsCapMode] in
+def C_CSD : CCheriStore_rri<0b111, "c.csd", GPRC, uimm8_lsb000> {
+  bits<8> imm;
+  let Inst{12-10} = imm{5-3};
+  let Inst{6-5} = imm{7-6};
+}
+
+let hasSideEffects = 0, mayLoad = 0, mayStore = 0, isCall = 1,
+    DecoderNamespace = "RISCV32CapModeOnly_", Defs = [C1],
+    Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV32, IsCapMode]  in
+def C_CJAL : RVInst16CJ<0b001, 0b01, (outs), (ins simm12_lsb0:$offset),
+                        "c.cjal", "$offset">;
+
+let hasSideEffects = 0, mayLoad = 0, mayStore = 0 in
+def C_CIncOffsetImm16CSP : RVInst16CI<0b011, 0b01, (outs CSP:$rd_wb),
+                                      (ins CSP:$rd, simm10_lsb0000nonzero:$imm),
+                                      "c.cincoffset16csp", "$rd, $imm"> {
+  let Constraints = "$rd = $rd_wb";
+  let Inst{12} = imm{9};
+  let Inst{11-7} = 2;
+  let Inst{6} = imm{4};
+  let Inst{5} = imm{6};
+  let Inst{4-3} = imm{8-7};
+  let Inst{2} = imm{5};
+}
+
+let DecoderNamespace = "RISCV32CapModeOnly_",
+    Predicates = [HasCheri, HasCheriRVC, HasStdExtC, HasStdExtD, IsRV32, IsCapMode] in
+def C_CFLDCSP : CCheriStackLoad<0b001, "c.cfldcsp", FPR64, uimm9_lsb000> {
+  let Inst{6-5} = imm{4-3};
```

Comment:
let Inst{12} = imm{5} ?

Reply:
Yep, seems I have indeed missed that

Reply:
Oh, no, I see why. Because every CI-format instruction has bit 5 of the immediate in bit 12 of the instruction, that's factored out into RVInst16CI's definition. And I'm pretty sure TableGen shouts at you when you've left bits of the encoding uninitialised, but can't say for certain.

Reply:
Interestingly stack stores (CSS-format) don't do the same thing, I guess because there's always at least one other bit that's contiguous so it doesn't save any lines of code as bit 12/5 can always be folded into an existing line in the instruction definition, but still a bit inconsistent.

Reply:
Okay, sounds good